test(subscriber): prefer `sleep` over `yield_now` in tests

hds opened the pull request:
A flakiness problem has been discovered with the `console-subscriber`
integration tests introduced in #452. Issue #473 is tracking the issue.

It has been observed that we only "miss" the wake operation event when
it comes from `yield_now()`, but not when it comes from a task that
yielded due to `sleep`, even when the duration is zero. it is likely
that this is due to nature of the underlying race condition.

This change removes all the calls to `yield_now()` from the `framework`
tests, except those where we wish to actually test self wakes.

hds committed:
test(subscriber): prefer `sleep` over `yield_now` in tests

A flakiness problem has been discovered with the `console-subscriber`
integration tests introduced in #452. Issue #473 is tracking the issue.

It has been observed that we only "miss" the wake operation event when
it comes from `yield_now()`, but not when it comes from a task that
yielded due to `sleep`, even when the duration is zero. it is likely
that this is due to nature of the underlying race condition.

This change removes all the calls to `yield_now()` from the `framework`
tests, except those where we wish to actually test self wakes.

hds committed:
Merge branch 'main' into hds/console-test-prefer-sleep

hawkw reviewed (approved):
overall, this looks good to me! 

i think it might be good if there were comments explaining the use of zero-duration sleeps rather than `yield_now`s, so that nobody comes along later and thinks "these seem like they could be replaced with `yield_now`", without the original context for why we use `sleep` here.

if you don't mind adding comments, let's get this merged!

hds committed:
clarify why we're using sleep instead of yield_now

Moved all the sleeps out into a separate function and use it to describe
why we're using sleep instead of yield_now when there's no difference.

hds commented:
Comments added!
